src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs

```diff
@@ -0,0 +1,476 @@
+            writeGenerators();
+            writer.WriteKey("emitOptions");
+            WriteEmitOptions(writer, emitOptions);
+
+            writer.WriteObjectEnd();
+
+            return builder.ToStringAndFree();
+
+            void writeAdditionalTexts()
+            {
+                writer.WriteArrayStart();
+                foreach (var additionalText in additionalTexts)
+                {
+                    writer.WriteObjectStart();
+                    WriteFileName(writer, "fileName", additionalText.Path, options);
+                    writer.WriteKey("text");
+                    WriteSourceText(writer, additionalText.GetText());
+                    writer.WriteObjectEnd();
+                }
+                writer.WriteArrayEnd();
+            }
+
+            void writeAnalyzers()
+            {
+                writer.WriteArrayStart();
+                foreach (var analyzer in analyzers)
+                {
+                    WriteType(writer, analyzer.GetType());
+                }
+                writer.WriteArrayEnd();
+            }
+
+            void writeGenerators()
+            {
+                writer.WriteArrayStart();
+                foreach (var generator in generators)
+                {
+                    WriteType(writer, generator.GetType());
+                }
+                writer.WriteArrayEnd();
+            }
+
+            void ensureNotDefault<T>(ref ImmutableArray<T> array)
+            {
+                if (array.IsDefault)
+                {
+                    array = ImmutableArray<T>.Empty;
+                }
+            }
+        }
+
+        internal string GetKey(EmitOptions? emitOptions)
+        {
+            var (writer, builder) = CreateWriter();
+            WriteEmitOptions(writer, emitOptions);
+            return builder.ToStringAndFree();
+        }
+
+        private void WriteCompilation(JsonWriter writer, Compilation compilation, DeterministicKeyOptions options)
+        {
+            writer.WriteObjectStart();
+            writeToolsVersions();
+
+            writer.WriteKey("options");
+            WriteCompilationOptions(writer, compilation.Options);
+
+            writer.WriteKey("syntaxTrees");
+            writer.WriteArrayStart();
+            foreach (var syntaxTree in compilation.SyntaxTrees)
+            {
+                WriteSyntaxTree(writer, syntaxTree, options);
+            }
+            writer.WriteArrayEnd();
+
+            writer.WriteKey("references");
+            writer.WriteArrayStart();
+            foreach (var reference in compilation.References)
+            {
+                WriteMetadataReference(writer, reference);
+            }
+            writer.WriteArrayEnd();
+            writer.WriteObjectEnd();
+
+            void writeToolsVersions()
+            {
+                writer.WriteKey("toolsVersions");
+                writer.WriteObjectStart();
+                if (0 != (options & DeterministicKeyOptions.IgnoreToolVersions))
```

Comment:
nit: constant 0 should be on the right. Unless we are making a style decision that we put the 0 on the left for these kinds of flag checks?